oauthenticator/okpy.py

```diff
@@ -0,0 +1,109 @@
+    _OAUTH_AUTHORIZE_URL =  "https://okpy.org/oauth/authorize"
+    def get(self):
+        self.authorize_redirect(
+            redirect_uri = self.authenticator.oauth_callback_url,
+            client_id = self.authenticator.client_id,
+            scope= ['all'],
+            response_type = 'code')
+
+class OkpyCallbackHandler(BaseHandler, OAuth2Mixin):
+    """ Basic handler for Okpy callback.
+        Calls authenticator to verify user and saves the state in user.
+    """
+    def get_auth_request(self, code):
+        params = dict(
+            redirect_uri = self.authenticator.oauth_callback_url,
+            code = code,
+            grant_type = 'authorization_code'
+        )
+        b64key = base64.b64encode(
+            bytes("{}:{}".format(self.authenticator.client_id,
+                                 self.authenticator.client_secret), "utf8"
+                 )
+        )
+        url = url_concat(OAUTH_ACCESS_TOKEN_URL, params)
+        req = HTTPRequest(url,
+                method = "POST",
+                headers = { "Accept": "application/json",
+                            "User-Agent": "JupyterHub",
+                            "Authorization": "Basic {}".format(b64key.decode("utf8"))},
+                            body = '' # Body is required for a POST...
+                          )
+        return req
+
+    @gen.coroutine
+    def get(self):
+        username, state = yield self.authenticator.authenticate(self, None)
+        if username:
+            user = self.user_from_username(username)
+            user.auth_state = state
+            self.db.add(user)
+            self.db.commit()
+            self.set_login_cookie(user)
+            self.redirect(url_path_join(self.hub.server.base_url, 'home'))
+        else:
+            raise web.HTTPError(403, "Authentication Failed.")
+
+    def get_user_info_request(self, access_token):
+        headers = {"Accept": "application/json",
+                   "User-Agent": "JupyterHub",
+                   "Authorization": "Bearer {}".format(access_token)}
+        params = {"envelope" : "false"}
+        url = url_concat(OKPY_USER_URL, params)
+        req = HTTPRequest(url, method = "GET", headers = headers)
+        return req
+
+class OkpyOAuthenticator(OAuthenticator, OAuth2Mixin):
+    login_service = "Okpy"
+    callback_handler = OkpyCallbackHandler
+    login_handler = OkpyLoginHandler
+    @gen.coroutine
+    def authenticate(self, handler, data = None):
+        code = handler.get_argument("code", False)
+        if not code:
+            raise web.HTTPError(400, "Authentication Cancelled.")
+        http_client = AsyncHTTPClient()
+        auth_request = handler.get_auth_request(code)
+        response = yield http_client.fetch(auth_request)
+        if not response:
+            self.clear_all_cookies()
+            raise HTTPError(500, 'Authentication Failed: Token Not Acquired')
+        state = json.loads(response.body.decode('utf8', 'replace'))
```

Comment:
This is the downside of auth_state, and why I don't actually like it very much and think maybe we should remove it. The Hub has a policy of never storing auth info in cleartext in the db, because if the db is compromised everything is a mess. If Authenticators start doing this, that is violated, so I don't want any Authenticators in this repo to set a bad example by doing so.

What's more important is the ultimate goal of passing a token (or certs, etc.) from the upstream auth service to the Spawner on start (not on login). Storing this token in auth_state isn't a particularly good idea because the lifecycle of the data is wrong. Really, `start` should be the action where the token is requested, not login. Keeping that data in memory gives it a better lifetime, but one problem is left unsolved: when to retrieve a new token for a user with a valid login cookie. This is still true when the token is stored in auth_state, though, because there is no check for whether the token is valid. So if the token is revoked, the behavior is the same as the only-in-memory case where the token is lost (but worse, because it's present and thinks it's valid).

What we really need is a mechanism to request a token on start - such as allow `pre_spawn_start()` to be a more complicated action that may involve OAuth redirects. I'm not sure that's feasible, though. What should be feasible and solve this problem is if the Authenticator could intercede and double-check "Is the auth for this user still valid" and trigger re-login, that would also completely remove the need for auth_state in any case I'm aware of.